docs/design/deployment.adoc

```diff
@@ -5,7 +5,35 @@ v0.1, 2020-05-17
 
 NOTE: This document is under development
 
-== Migration Plane from Openstack Neutron
+== Migration Plan from Openstack Neutron
+
+First, the structures of Neutron models are not consistent with the Alcor's. 
+We need a tool that can transform the existing Neutron network configurations into the Alcor's, so that Alcor can process it.
+These converted network configurations should be handled by Alcor as batch creation operation messages with all parameters specified.
+
+However, we still want users to be able to continue configuring on this openstack environment, 
+as we use the tool to convert the netruon network configurations to Alcors.
+At this time, the virtual switch in compute node side should still be taken over by neutron agents, 
+and an alcor control agent should be in a state that can normally receive configuration, 
+but does not send configuration to the virtual switch.
+The newly generated configuration during this conversion may not be simultaneously converted to Alcor and processed, 
+therefore, we also need to intercept all the operation messages received by the Neutron during the conversion period.
+And in addition to passing them through to the Neutron so that the Neutron can normally process the messages,
+we also need to copy those operation messages and temporarily cache them.
+
+image::images/neutron_migration_phase1.png["Migration Phase1", width=1024, link="images/neutron_migration_phase1.png"]
+
+After the conversion of the original network configurations is completed, 
+these cached operation messages should also be converted to alcor's messages and processed.
+
+image::images/neutron_migration_phase2.png["Migration Phase2", width=1024, link="images/neutron_migration_phase2.png"]
+
+After that, the openstack CLI can switch to send operation messages directly to Alcor.
```

Comment:
This OpenStack CLI changes fall into one question: are we capable of asking all clients of the existing in-house OpenStack clusters upgrade with new pointer to Alcor? It might or may not. 

The existing Neutron APIs could be called by clients in house (like Nova) and in customer's sides (like deployment scripts). 
* Easier for the in-house clients 
* As we don't have control customer's clients, it would be very tricky here. One possible option is use to same DNS name pointing to different backend IP addresses. But need to handle with long DNS expiration. 

Let us have more discussion on this.

Reply:
In public cloud scenario, a provider is more likely to use a API gateway mechanism. And those APIs for users would be regitested there, each with a mapping from external URI to internal URI. So if there is any conversion on the internal URI side, we only need to override the mapping to the new internal URI in the API gateway. Those changes may not be awared by users, because the DNS name and URI for a API is not changed.
For the in-house clients, there is a similiar mechanism, but the adoption is only a suggestion, so those in-house clients might need a conversion step.

Reply:
> In public cloud scenario, a provider is more likely to use a API gateway mechanism. And those APIs for users would be regitested there, each with a mapping from external URI to internal URI. So if there is any conversion on the internal URI side, we only need to override the mapping to the new internal URI in the API gateway. Those changes may not be awared by users, because the DNS name and URI for a API is not changed.

Cool. In the public cloud scenario, this means the biggest client of Alcor is likely the in-house API Gateway so transition is not a big issue then. 

> For the in-house clients, there is a similar mechanism, but the adoption is only a suggestion, so those in-house clients might need a conversion step.

Got you.